### PR TITLE
Add insert-next commands (loadfile, loadlist, drag-and-drop)

### PR DESCRIPTION
**Pull request by DavidRV00**
Hello, sorry if this is a little large to just send without prior discussion. Please let me know if it could use alterations, refactoring, stylistic improvements, etc, as well as any feedback on the general utility of it.

--------------

This change adds two new commands (`insert-next` and `insert-next-play`) which mirror the existing commands, `append` and `append-play` in functionality, with the difference that they insert directly after the current playlist entry, rather than at the end of the playlist.

This change gives MPV a piece of functionality already found in (for example) Spotify's media player: "play next". Additionally, using the new `insert-next` command, users can trivially write a script to play a new piece of media immediately without otherwise clearing or altering the remainder of the playlist.

My motivation for making this change is the following use-case: I have a playlist already open and playing, then I come across a piece of media I want to play *right now*, but I don't want to clear the existing playlist. AFAIK, there is not currently a way to do this with any reasonable sequence of commands. Adding `insert-next` allows for it.

I have tested the new feature in all the ways I could think of (commands via ipc messages, drag-and-drop), as well as making sure replace and append are unbroken. I've only tested on X11; the wayland-specific stuff I have not been able to try myself.

### CODE REVIEW

**Comment by Dudemanguy** [comment]
It might be worth it to generalize the underlying implementation so you can insert the new file in any position in the playlist (like `insert-at` and an additional index argument). Of course, `insert-next` is a sensible shorthand like `append`.

**Comment by DavidRV00** [comment]
I could do that. Then probably it would make sense for even `append` to be implemented in terms of the same "insert-at" logic, too.

**Comment by Dudemanguy** [comment]
That's what I had in mind too.

**Comment by DavidRV00** [comment]
Sounds good. I'll do that refactor and send up the commit later today or tomorrow.

**Comment by DavidRV00** [comment]
Alright I've done that. Replaced `playlist_insert_next` with `playlist_insert_at`, and removed `playlist_add` completely in favor of just using the new function in its place.

There was a `playlist_add_file` function used in a few places that does some file-related stuff on top of the appending, which I've renamed to `playlist_append_file`, but otherwise left unchanged since it's a little more involved than just playlist operations.

If this looks good so far, I'll go ahead and add the `insert-at` command, and simplify the implementation of `loadlist` using it. But fyi I'm not planning to make any additional changes to drag-n-drop, because I can't imagine anybody ever wanting to `insert-at` with drag-n-drop.

**Comment by Dudemanguy** [comment]
>But fyi I'm not planning to add the insert-at option itself to drag-n-drop, because I can't imagine anybody ever wanting to insert-at with drag-n-drop.

Yeah that's perfectly fine. There's no real practical way to handle that particular case.

File: common/playlist.c [hunk 1]
```diff
@@ -60,13 +60,22 @@ static void playlist_update_indexes(struct playlist *pl, int start, int end)
         pl->entries[n]->pl_index = n;
 }
 
-void playlist_add(struct playlist *pl, struct playlist_entry *add)
+// Inserts the entry so that it comes directly before "at" (or append to end, if at==NULL).
+void playlist_insert_at(struct playlist *pl, struct playlist_entry *add,
+                        struct playlist_entry *at)
```
[the remark this answers is not in the document]
Comment (DavidRV00) [reply common/playlist.c:65]: Oh sorry, it *does* insert *at* the "at" position. What I'm trying to say in the comment is that it inserts before the element which is currently located at the "at" position (which is equivalent, of course). I'll clarify the comment.

[the remark this answers is not in the document]
Comment (Dudemanguy) [reply common/playlist.c]: Ah cool. Yeah I interpreted that the wrong way at first.

**Comment by DavidRV00** [comment]
Hey @Dudemanguy I could actually use some advice for the `insert-at` command before I continue with it. I'm not sure where / in what value I should take in the new "at" parameter.

Currently the format of the command is `loadfile <url> [<flags> [<options>]]`, where the `options` just appear to be playback options, so I don't know if it fits to put the "at" value in there. Do you think we should add a new optional "at" value as the new 3rd argument (ie, before `options`)? Or, maybe should it just be included in the `flags` argument (ie, like `loadfile ... insert-at=5` ) ? Do we currently have any good examples of this kind of thing?

**Comment by Dudemanguy** [comment]
Sorry a bit late here. I think the best option is for it be included in the `flags` so that it doesn't break old commands and also keeps the amount of arguments the same for every possible flag. I guess this means that you'll probably have to make the flags argument a string internally instead of an integer so you can parse out the position of the `insert-at`, but that's probably OK.

**Comment by DavidRV00** [comment]
Great, thanks.  I think that'll work. I should be able to finish this up sometime in the next few days.

**Comment by DavidRV00** [comment]
Sorry pushed the wrong branch. I'll fix that real quick.

**Comment by Dudemanguy** [comment]
I apologize. I know I'm contradicting what I said earlier and have essentially wasted a bunch of your time, but thinking about it more, all this complexity to avoid a minor change in the API command is just not really worth it. I doubt very many people even set `options` and I just realized that `loadlist` doesn't even have it so you can easily add a trailing argument there for free. So in short, I'm suggesting to just go with this:

```
diff --git a/player/command.c b/player/command.c
index e81874010b..2f8414318d 100644
--- a/player/command.c
+++ b/player/command.c
@@ -6669,6 +6669,7 @@ const struct mp_cmd_def mp_cmds[] = {
                 {"append", 1},
                 {"append-play", 2}),
                 .flags = MP_CMD_OPT_ARG},
+            {"index", OPT_INT(v.i), OPTDEF_INT(-1)},
             {"options", OPT_KEYVALUELIST(v.str_list), .flags = MP_CMD_OPT_ARG},
         },
     },
@@ -6680,6 +6681,7 @@ const struct mp_cmd_def mp_cmds[] = {
                 {"append", 1},
                 {"append-play", 2}),
                 .flags = MP_CMD_OPT_ARG},
+            {"index", OPT_INT(v.i), OPTDEF_INT(-1)},
         },
         .spawn_thread = true,
         .can_abort = true,
```

i.e. an additional argument for the index that of course only has an effect with `insert-at` (I left out the other flags you added for brevity.

**Comment by DavidRV00** [comment]
No worries, that's fair. It wasn't clear from the outset how much extra stuff it would take to make it work. If you're fine with the extra parameter in the command, then I'll go ahead and do this. Should be able to send it up pretty soon.

**Comment by DavidRV00** [comment]
There it is; I've reverted most of the previous commit and used the int options and index arg insteads. Let me know how that looks.

File: player/command.c [hunk 2]
```diff
@@ -5489,29 +5501,78 @@ static void cmd_expand_path(void *p)
...
 
+static struct load_action get_load_action(struct MPContext *mpctx, int action_flag)
+{
+    switch (action_flag) {
+    case 0: { // replace
+        return (struct load_action){LOAD_TYPE_REPLACE, .play = true};
+    }
+    case 1: { // append
+        return (struct load_action){LOAD_TYPE_APPEND, .play = false};
+    }
```
Comment (Dudemanguy) [inline player/command.c:5512]:
```suggestion
    case 0: // replace
        return (struct load_action){LOAD_TYPE_REPLACE, .play = true};
    case 1: // append
        return (struct load_action){LOAD_TYPE_APPEND, .play = false};
```

Could nuke all the extra parenthesis here. Ditto for the rest of it. More concise and easier to read imo.

**Review by Dudemanguy — COMMENTED** [review]
Changes look overall good though and seem to work for me with some testing. Just start squashing some commits I would say. Also, if you could note the change in `DOCS/interface-changes.rst`, that would be great. Particularly about the order of the arguments of the `loadfile` command changing a bit.

File: player/command.c (hunk 2 above)
Comment (DavidRV00) [reply player/command.c:5512]: Gotcha, will do.

**Comment by DavidRV00** [comment]
For sure, I'll edit interface-changes.rst, squash some commits, then force push that up here shortly.

**Comment by DavidRV00** [comment]
Alright, there we go, and merge conflict in the docs resolved. Let me know if it could use any more squashing or cleanup.

**Comment by Dudemanguy** [comment]
Thanks, just one final thing about the commit subjects. I think you can just shorten the prefix to `player:` and `input:`

**Comment by DavidRV00** [comment]
Thanks, done.

**Review by Dudemanguy — APPROVED** [review]
Thanks for doing all of this despite me greatly expanding the scope the patch!

**Comment by DavidRV00** [comment]
No worries, happy to do it!

**Pull request merged**